lib/core/test/unit/Cardano/Wallet/Primitive/Types/Tx/TxSeqSpec.hs

```diff
@@ -0,0 +1,532 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{- HLINT ignore "Hoist not" -}
+
+module Cardano.Wallet.Primitive.Types.Tx.TxSeqSpec
+    ( spec
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Primitive.Types.Address.Gen
+    ( genAddress )
+import Cardano.Wallet.Primitive.Types.Hash
+    ( Hash (..) )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( AssetId )
+import Cardano.Wallet.Primitive.Types.TokenMap.Gen
+    ( genAssetId, shrinkAssetId )
+import Cardano.Wallet.Primitive.Types.TokenPolicy
+    ( TokenName (..), TokenPolicyId (..) )
+import Cardano.Wallet.Primitive.Types.Tx.TxSeq.Gen
+    ( ShrinkableTxSeq, genTxSeq, getShrinkPhase, getTxSeq, shrinkTxSeq )
+import Cardano.Wallet.Primitive.Types.UTxO.Gen
+    ( genUTxO )
+import Data.Function
+    ( (&) )
+import Data.Function.Utils
+    ( isInjectiveOver )
+import Data.Maybe
+    ( fromMaybe, mapMaybe )
+import Test.Hspec
+    ( Spec, describe, it )
+import Test.QuickCheck
+    ( Arbitrary (..)
+    , CoArbitrary (..)
+    , Fun (..)
+    , Function (..)
+    , Property
+    , Testable (..)
+    , applyFun
+    , checkCoverage
+    , chooseInt
+    , cover
+    , forAll
+    , property
+    , withMaxSuccess
+    , (===)
+    , (==>)
+    )
+import Test.QuickCheck.Extra
+    ( ScaleDiv (..), genShrinkSequence, labelInterval, shrinkWhile )
+import Test.QuickCheck.Instances.ByteString
+    ()
+
+import qualified Cardano.Wallet.Primitive.Types.Tx.TxSeq as TxSeq
+import qualified Data.Foldable as F
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Set as Set
+
+spec :: Spec
+spec = do
+
+    describe "Generation" $ do
+
+        describe "genTxSeq" $ do
+            it "prop_genTxSeq_isValid" $
+                prop_genTxSeq_isValid
```

Comment:
The main justification for repeating the property name in the description is that when a property fails, it's very easy to search for the property function name and locate it in the codebase. Unfortunately, `hspec` doesn't print the property name by default when there's a failure.

I actually don't think that these description fields are the right place to describe what the property does. Very often, there's no easy way to write a concise description that fits on one line. Putting it on multiple lines often requires the use of awkward line continuation characters.

I do think that if we need to provide a natural language description of the property, then we'd derive more benefit from putting that description in a **_comment_** attached to the property definition itself. Comments have the advantage of allowing many lines, and code examples.

If you think there are any properties that are currently unclear and would benefit from comments, I'd be happy to add them. Let me know!

In an ideal world, I think we would just generate these `hspec` structures automatically, instead of hand-coding them.